TST:add test for df replace GH34871

nwweber opened the pull request:
- [x] improves upon #34871
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry <---- is that helpful for this specific regression test? happy to write this if you think this helps! :)

add tiny regression test for not failing on calling df.replace() with Period column

working on this revealed other quirks of df.replace(): it now does not raise an exception anymore, and the dataframe returned by replace() has the right values. However, the dtype of a Period column changes to Object, which is surprising. 

I've added the observations to GH34871

nwweber committed:
initial version of regression test for GH#34871

nwweber committed:
reduce test to only fail for exception

test now does not check for actual equality between original and new dataframe anymore, this behavior needs treduce test to only fail for exception

test now does not check for actual equality between original and new dataframe anymore, this behavior needs to be addressed in another bug fix be addressed in another bug fix

jreback commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1403,3 +1403,28 @@ def test_replace_with_duplicate_columns(self, replacement):
+    def test_replace_period_ignore_float(self):
+        """
+        Regression test for GH#34871: if df.replace(1.0, 0.0) is called on a df
+        with a Period column the old, faulty behavior is to raise TypeError
```
you don't need to reference older versions like this, this commit would only be on master.

nwweber replied:
thanks for the advise, removed that part!

jreback commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1403,3 +1403,28 @@ def test_replace_with_duplicate_columns(self, replacement):
+        test that triggers for the old, broken behavior, i.e.
+        TypeError is raised
+        """
+        df = pd.DataFrame({"Per": [pd.Period("2020-01")] * 3})
```
please use explicit comparisons, if this is *only* a test which is expected to fail then use ```@pytest.xfail(reason=....)```

jreback replied:
IOW don't comment out code

nwweber replied:
Hey! It's sort of the opposite, I expect being able to call df.replace(...) without it raising an exception for these specific inputs.

Would I then simply run "df.replace(...)" in the unit test, and that's sufficient?

jreback replied:
you would need to assert the results

nwweber replied:
updated, code now asserts expected result, also removed commented-out blocks of code

nwweber committed:
adress change request for PR

don't reference old version
assert expected result

jreback commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1403,3 +1403,17 @@ def test_replace_with_duplicate_columns(self, replacement):
+        df = pd.DataFrame({"Per": [pd.Period("2020-01")] * 3})
+        df_after_replace = df.replace(1.0, 0.0)
+
+        df_expected_result = pd.DataFrame({"Per": [pd.Period("2020-01")] * 3})
```
can you call things result= and expected= its much easier to read (and doesn't mix in the same name)

nwweber replied:
done!

jreback commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1403,3 +1403,17 @@ def test_replace_with_duplicate_columns(self, replacement):
+
+        df_expected_result = pd.DataFrame({"Per": [pd.Period("2020-01")] * 3})
+        # currently replace() changes dtype from Period to object
+        df_expected_result["Per"] = df_expected_result["Per"].astype(object)
```
so this is wrong, we want to xfail this test, this *should* not change the dtype, pls add the issue number in the xfail itself

jreback replied:
meaning, yes it doesn't raise an expection, but the output is still converted to object

nwweber replied:
gotcha, thanks for teaching me about xfail. i changed the test to now expect the result we would like to see from replace() in the future, and to xfail for now.

nwweber committed:
test for desired behavior and xfail for now

unit test for replace() now checks whether the dataframe returned is equal to how we want the function to behave, i.e. no changes to either data values or data type. this behavior needs to be implemented in the future, thux xfail-ing for now.

jreback commented:
thanks @nwweber 

happy for you to tackle https://github.com/pandas-dev/pandas/issues/34871 :->

nwweber commented:
> thanks @nwweber
> 
> happy for you to tackle #34871 :->

i'll have a look over the weekend :)
